common/src/Common/Network.hs

```diff
@@ -208,3 +226,82 @@ renderNodeRef (NodeRef (URI.Authority mUser h mp)) =
+compileCode :: Text -> Either String [Term Name]
+compileCode = first show . compileExps mkEmptyInfo <=< parseExprs
+
+-- | Parse the balance checking object into a map of account balance changes and
+-- the result from the inner code
+parseWrappedBalanceChecks :: PactValue -> Either Text (Map AccountName Decimal, PactValue)
+parseWrappedBalanceChecks = first ("parseWrappedBalanceChecks: " <>) . \case
+  (PObject (ObjectMap obj)) -> do
+    let lookupErr k = case Map.lookup (FieldKey k) obj of
+          Nothing -> Left $ "Missing key '" <> k <> "' in map: " <> renderCompactText (ObjectMap obj)
+          Just v -> pure v
+    before <- parseAccountBalances =<< lookupErr "before"
+    result <- parseResults =<< lookupErr "results"
+    after <- parseAccountBalances =<< lookupErr "after"
+    pure (Map.unionWith subtract before after, result)
+  v -> Left $ "Unexpected PactValue (expected object): " <> renderCompactText v
+
+-- | Turn the object of account->balance into a map
+parseAccountBalances :: PactValue -> Either Text (Map AccountName Decimal)
+parseAccountBalances = first ("parseAccountBalances: " <>) . \case
+  (PObject (ObjectMap obj)) -> do
+    m <- for (Map.toAscList obj) $ \(FieldKey accountText, pv) -> do
+      bal <- case pv of
+        PLiteral (LDecimal d) -> pure d
+        t -> Left $ "Unexpected PactValue (expected decimal): " <> renderCompactText t
+      acc <- mkAccountName accountText
+      pure (acc, bal)
+    pure $ Map.fromList m
+  v -> Left $ "Unexpected PactValue (expected object): " <> renderCompactText v
+
+-- | Get the last result, as we would under unwrapped deployment.
+parseResults :: PactValue -> Either Text PactValue
+parseResults = first ("parseResults: " <>) . \case
+  PList vec -> pure $ V.last vec
+  v -> Left $ "Unexpected PactValue (expected list): " <> renderCompactText v
+
+-- | Wrap the code with a let binding to get the balances of the given accounts
+-- before and after executing the code.
+wrapWithBalanceChecks :: Set AccountName -> Text -> Either String Text
+wrapWithBalanceChecks accounts code = wrapped <$ compileCode code
+  where
+    getBalance :: AccountName -> (FieldKey, Term Name)
+    getBalance acc = (FieldKey (unAccountName acc), TApp
+      { _tApp = App
+        { _appFun = TVar (QName $ QualifiedName "coin" "get-balance" def) def
+        , _appArgs = [TLiteral (LString $ unAccountName acc) def]
+        , _appInfo = def
+        }
+      , _tInfo = def
+      })
+    -- Produce an object from account names to account balances
+    accountBalances = TObject
+      { _tObject = Pact.Types.Term.Object
+        { _oObject = ObjectMap $ Map.fromList $ map getBalance $ Set.toAscList accounts
+        , _oObjectType = TyPrim TyDecimal
+        , _oKeyOrder = Nothing
+        , _oInfo = def
+        }
+      , _tInfo = def
+      }
+    -- It would be nice to parse and compile the code and shove it into a
+    -- giant 'Term' so we can serialise it, but 'pretty' is not guaranteed to
```

Comment:
Not an issue for us/this release, but should probably pass this back to Kadena as it'd be nice to have some TH like `[pact| (foo "biscuit")|]` that would give us compile errors if given an invalid pact expression.